### PR TITLE
Notebook crud

### PR DESCRIPTION
**Pull request by zarsky-broad**
This is a big one. Sorry about that.

### CODE REVIEW

File: src/pages/workspaces/workspace/Notebooks.js [hunk 1]
```diff
@@ -1,14 +1,146 @@
...
+              printName, namespace, bucketName, destroyOld: false,
+              onDismiss: () => this.setState({ copyingNotebook: false }),
+              onSuccess: () => reloadList()
+            })
+          ],
+          [
+            deletingNotebook,
+            () => NotebookDeleter({
+              printName, namespace, bucketName,
+              onDismiss: () => this.setState({ copyingNotebook: false }),
```
Comment (panentheos) [inline src/pages/workspaces/workspace/Notebooks.js]: Bug: It looks like this should be `deletingNotebook`

File: package.json [hunk 2]
```diff
@@ -3,18 +3,19 @@
   "version": "0.0.1",
   "private": true,
   "dependencies": {
-    "@clr/icons": "^0.11.12",
+    "@clr/icons": "^0.11.14",
     "@webcomponents/custom-elements": "^1.1.0",
-    "lodash": "^4.17.5",
-    "rc-table": "^6.1.7",
-    "react": "^16.3.1",
+    "lodash": "^4.17.10",
```
Comment (panentheos) [inline package.json:8]: Discussion: Did all these need to be upgraded for this feature? If not, it might be a bit safer to split upgrades out into their own PR so we can test and review them separately.

File: src/pages/workspaces/workspace/Container.js [hunk 3]
```diff
@@ -48,7 +48,7 @@ const tabComponents = {
  * @param {string} name
  * @param {string} [activeTab]
  */
-export const WorkspaceContainer = hh(class WorkspaceContainer extends Component {
+export class WorkspaceContainer extends Component {
```
Comment (panentheos) [inline src/pages/workspaces/workspace/Container.js:51]: Discussion: We should probably decide on a consistent approach for exporting. I would vote for always exporting the React class directly, since exporting the "factory" function is nonstandard and has already caused us to make other unusual accommodations (e.g. in the tests).

File: src/components/notebook-utils.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    ])
+  }
+})
+
+export const NotebookDuplicator = hh(class NotebookRenamer extends Component {
+  render() {
+    const { destroyOld, printName, namespace, bucketName, onDismiss, onSuccess } = this.props
+    const { newName, processing, failure } = this.state
+
+    return Modal({
```
Comment (panentheos) [inline src/components/notebook-utils.js:133]: Discussion: Given the number of parallel conditionals here, this feels like it might want to be two code paths, with the common stuff factored out. (Just an instinct, happy to be convinced otherwise)

**Review by panentheos — APPROVED** [review]
Looks great! See one bug comment before merging (the rest are non-critical discussion items)

File: src/components/ShowOnClick.js [hunk 5]
```diff
@@ -25,10 +25,11 @@ export default hh(class ShowOnClick extends Component {
 
   render() {
     const { visible } = this.state
-    const { containerProps, button, bgProps, closeOnClick = true, closeOnEsc = true, children } = this.props
+    const { containerProps, button, disabled, bgProps, closeOnClick = true, closeOnEsc = true, children } = this.props
 
     return div(_.merge({
-        onClick: (e) => {
+        style: visible ? { zIndex: 99, position: 'relative' } : undefined,
```
Comment (dmohs) [inline src/components/ShowOnClick.js]: What is the significance of 99 here? Would it be worth naming this value via a constant to add clarity?

File: src/pages/workspaces/workspace/Notebooks.js [hunk 6]
```diff
@@ -1,14 +1,146 @@
...
+  render() {
+    const { namespace, name, updated, listView, notebookAccess, bucketName, clusterUrl, wsName, reloadList } = this.props
+    const { renamingNotebook, copyingNotebook, deletingNotebook } = this.state
+    const printName = name.slice(10, -6) // removes 'notebooks/' and the .ipynb suffix
+
+    const hideMenu = () => this.notebookMenu.setVisibility(false)
+
+    const notebookMenu = ShowOnClick({
+        ref: instance => this.notebookMenu = instance,
+        disabled: !notebookAccess,
```
Comment (dmohs) [inline src/pages/workspaces/workspace/Notebooks.js:24]: When I click on a disabled card, I get no indication of why it is disabled. Might be worth filing a bug and addressing later.

**Review by dmohs — APPROVED** [review]

File: src/pages/workspaces/workspace/Container.js (hunk 3 above)
Comment (MattPutnam) [reply src/pages/workspaces/workspace/Container.js:51]: I agree

**Review by MattPutnam — APPROVED** [review]

File: package.json (hunk 2 above)
Comment (zarsky-broad) [reply package.json:8]: You're absolutely right; I wanted to update all the react stuff, since 16.3 is pretty recent and I wanted bugfixes for that (I made sure to look at patch notes for everything), but lodash and icons I just didn't bother to ignore.

File: src/components/ShowOnClick.js (hunk 5 above)
Comment (zarsky-broad) [reply src/components/ShowOnClick.js]: It's purpose is purely to make sure that element is high enough, it's really a magic number. I can totally make it a constant.

File: src/components/notebook-utils.js (hunk 4 above)
Comment (zarsky-broad) [reply src/components/notebook-utils.js:133]: There are really only 4 places where the split happens, and 3 of them are just string labels. Since the rename and copy methods take the same arguments, I thought it made sense to combine the modals.

File: src/pages/workspaces/workspace/Container.js (hunk 3 above)
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/Container.js:51]: So, this is potentially complicated.

If we always just export classes, we'll **always** have to wrap them in `h()` when we call them.

The reason they're not `hh()` wrapped here is that it breaks the hot reload; the only components that aren't `hh()` wrapped are those that are passed into a `Nav.defPath`.
